### PR TITLE
NEW Add default filters to data grid and flexibility to Query filter conditions

### LINKED ISSUE #141: Column results conditions reflected in front-end
Built in #23, this needs to be tied to pre-filter data populated into the CKAN registry table by default.

It should be be used to update the select statement, using the CKAN SQL endpoint, which is updated with a javascript function.

### LINKED ISSUE #156: Result Condition field match type doesn't propagate to hidden input until changed
* Edit a column
* Don't change the Result conditions match type select
* Enter "Auckland" into the Result conditions text field
* Inspect the value of the hidden input (`jQuery('input[name="DisplayConditions"]').val()`)

Expected: `"{"0":{"match-select":"1","match-text":"Auckland"}}"`

Actual: `"{"0":{"match-text":"Auckland"}}"`

This is important for the frontend.

### PR DESCRIPTION
**Pull request by robbieaverill**
This change:

* allows Query to accept filters that are optionally strict (doesn't add `%` character wrapping around the values to filter against) and allow match/don't match logic in these filters
* converts ResourceField boolean fields to be expected and handled as booleans rather than having various exceptions to allows proptypes
* ensures that ResultConditions data is decoded from JSON in getCKANClientConfig
* fixes some bugs in ResutConditionsField which was preventing it from setting its initial state correctly (fixed #156)

Resolves #141

### CODE REVIEW

File: client/src/components/CKANRegistryDisplay.js [hunk 1]
```diff
@@ -117,25 +124,54 @@ class CKANRegistryDisplay extends Component {
...
+        // any configured display conditions defined
+        return field.DisplayConditions.length && field.ShowInDetailView;
+      })
+      .forEach((field) => {
+        field.DisplayConditions.forEach((displayCondition) => {
+          // Add our filter statement
+          query.filter(
+            field.OriginalLabel, // column
+            displayCondition['match-text'], // term
+            true, // strict
```
Comment (ScopeyNZ) [inline client/src/components/CKANRegistryDisplay.js]: Hmmmn. I wonder if this is what's expected... Should the "Must match" configuration on a column look for exact matches @camfindlay?

Comment (camfindlay) [reply client/src/components/CKANRegistryDisplay.js]: It's probably more a contains, the key use case for this is pre-filtering by some sort of category. In the case a column contains more than 1 category (eg. comma separated "Health, Youth") and let's say you wanted to create a "Youth services" register from a larger master dataset. You'd want to "Match "Youth" in the Category column. If it's an exact match, the above record wont be included even though it's a legit entry for our case "Youth != Heath, Youth". 

How is the dropdown filter working for delimiter separated terms? I'd imagine the pre-filter would work in much the same way.

Comment (robbieaverill) [reply client/src/components/CKANRegistryDisplay.js]: I’m thinking about it again - “must match Auckland” could also return South Auckland too. Maybe this isn’t required after all

**Comment by ScopeyNZ** [comment]
Ok I added a commit to switch back to partial matching. While I had it checked out I had a little play. I am getting a weird situation when I save a column without touching the "must match" filter. Using the [family-services-directory] (https://catalogue.data.govt.nz/dataset/family-services-directory) dataset I updated `PROVIDER_WEBSITE_1` changing only the displayed label. I then added a filter on `LEVEL_1_CATEGORY`. The resulting query that was generated was:
```
SELECT DISTINCT "PROVIDER_NAME", "PROVIDER_WEBSITE_1", "PHYSICAL_REGION", "_id" FROM "35de6bf8-b254-4025-89f5-da9eb6adf9a0"  WHERE ("PROVIDER_WEBSITE_1" NOT ILIKE '%undefined%') AND ("LEVEL_1_CATEGORY" NOT ILIKE '%Youth%') LIMIT 30 OFFSET 0
```

Should be a fairly simple fix - I'll see if I can get around to it this afternoon.

File: client/src/components/CKANRegistryDisplay.js [hunk 2]
```diff
@@ -117,25 +124,54 @@ class CKANRegistryDisplay extends Component {
...
 
+  /**
+   * Finds all visible fields that have configured "ResultConditions", e.g. fields
+   * that should only show data in them that match certain conditions. When found,
+   * these will be applied as query filters before the data is loaded.
+   *
+   * @param {Query} query
+   */
+  applyDefaultFilters(query) {
+    const { fields } = this.props;
```
Comment (ScopeyNZ) [inline client/src/components/CKANRegistryDisplay.js]: This method is can have the call trace: `applyDefaultFilters <- resetQueryFilters <- constructor`. Is accessing props going to be a problem here? I can't remember whether it's safe to access props like this in the constructor in React components.

Comment (ScopeyNZ) [reply client/src/components/CKANRegistryDisplay.js]: Did my research. It's fine as long as you `super(props)` first. Probably should've guessed that.

**Comment by ScopeyNZ** [comment]
Okay I've added a commit fixing the issue I raised.

File: client/src/components/CKANRegistryDisplay.js [hunk 3]
```diff
@@ -422,10 +462,10 @@ CKANRegistryDisplay.propTypes = {
   fields: PropTypes.arrayOf(PropTypes.shape({
     OriginalLabel: PropTypes.string,
     ReadableLabel: PropTypes.string,
-    ShowInResultsView: PropTypes.oneOf([1, 0, true, false]),
-    ShowInDetailView: PropTypes.oneOf([1, 0, true, false]),
+    ShowInResultsView: PropTypes.bool,
+    ShowInDetailView: PropTypes.bool,
```
Comment (NightJar) [inline client/src/components/CKANRegistryDisplay.js]: 👍

**Pull request merged**